stingray/simulator/events.py

```diff
@@ -174,3 +177,48 @@ def gen_lc_from_events(event_list, bin_time, start_time=None,
         times = times + bin_time / 2.
 
     return times, lc.astype(np.float)
+
+def assign_energies(event_list, spectrum):
+
+    '''
+    Assign energies to event lists.
+
+    Parameters
+    ----------
+    event_list: array-like
+        Times of arrival of events
+    spectrum: 2-d array
+        Energies versus corresponding fluxes
+
+    Returns
+    -------
+    assigned_energies: array-like
+        Energies assigned to all events in event list
+    '''
+
+    assert spectrum.shape[0] != 2, "Spectrum must be a 2-d array."
+
+    assert np.all(np.isfinite(event_list)), "There are inf or NaN values in " \
+                                          "your event list!"
+
+    # Cast inputs as numpy arrays
+    event_list = np.array(event_list)
+
+    # Normalize spectrum by calculating cumulative density
+    fluxes = norm.cdf(np.array(spectrum[0]))
+    energies = np.array(spectrum[1])
+
+    lower, upper, mu, sigma = 0, 1, 0, 1
+
+    # Create a truncated normal distribution
+    X = truncnorm((lower - mu) / sigma, (upper - mu) / sigma, loc=mu, scale=sigma)
+
+    # Generate N random numbers between 0 and 1, where N is the size of event list
+    N = X.norm(size = len(event_list))
+
+    assigned_energies = np.array([])
+
+    for n in N:
+        assigned_energies = np.append(assigned_energies, energies[max(np.where([fluxes==n]))])
```

Comment:
This `append` operation is very slow. Since you know already how many photons you are going to generate, you can just preallocate `assigned_energies` to the right size